### ISSUE TITLE
TypeError: Converting circular structure to JSON

### ISSUE THREAD

**Issue by eatonjl** [artemrudenko/playwright-teamcity-reporter#10]
I am pretty new to this, so I am not sure how simple of an issue is with this. Perhaps it is more user error than code error. I am using Playwright 1.22 and I have two test files and they contain groupings inside of them. When I try to run the test, I get the following error (first part of each file path has been removed for this post): 
TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    |     property 'projects' -> object with constructor 'Array'
    |     index 0 -> object with constructor 'Object'
    --- property '_fullConfig' closes the circle
    at JSON.stringify (<anonymous>)
    at TeamcityReporter.onBegin (**[filepath]**/node_modules/playwright-teamcity-reporter/src/teamcity.reporter.ts:37:44)
    at Multiplexer.onBegin (**[filepath]**/node_modules/@playwright/test/lib/reporters/multiplexer.js:37:116)
    at Runner._runFiles (**[filepath]**/node_modules/@playwright/test/lib/runner.js:383:153)
    at Runner._run (**[filepath]**/node_modules/@playwright/test/lib/runner.js:224:12)
    at TimeoutRunner.run (**[filepath]**/node_modules/playwright-core/lib/utils/timeoutRunner.js:53:14)
    at raceAgainstTimeout (/**[filepath]**/node_modules/playwright-core/lib/utils/timeoutRunner.js:113:15)
    at Runner.runAllTests (**[filepath]**/node_modules/@playwright/test/lib/runner.js:182:20)
    at runTests (/**[filepath]**/node_modules/@playwright/test/lib/cli.js:162:18)
    at Ai.<anonymous> (/**[filepath]**/node_modules/@playwright/test/lib/cli.js:70:7)

Any idea what would be causing this?

**Comment by eatonjl** [comment]
I created a brand new Playwright install (which comes with an example test) and ran it to make sure it passed. Then I installed and tried to run with the reporter and received the same issue.

**Comment by artemrudenko** [comment]
> @artemrudenko Thank you! That seems to have fixed it. I do now see an error when the tests run. The error is this: "The service message is invalid because it does not end with ] character: ##teamcity[message text='{" Should I create a new issue for that one?

I have tested it with default playwright.config and all was ok. Could you please provide your config to check with it?

**Comment by eatonjl** [comment]
Sure thing!
import type { PlaywrightTestConfig } from '@playwright/test';
import { devices } from '@playwright/test';
import dotenv from 'dotenv';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
 dotenv.config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */
const config: PlaywrightTestConfig = {
  testDir: './tests',
  /* Maximum time one test can run for. */
  timeout: 30 * 1000,
  expect: {
    /**
     * Maximum time expect() should wait for the condition to be met.
     * For example in `await expect(locator).toHaveText();`
     */
    timeout: 10000
  },
  // Tell all tests to load signed-in state from 'storageState.json'.
  globalSetup: require.resolve('./global-setup'),
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: 1,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  //reporter: process.env.CI ? 'dot' : 'list',
  reporter: [
    ['playwright-teamcity-reporter', {'testMetadataArtifacts': 'test-results'}],
  ],
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    ignoreHTTPSErrors: true,
    storageState: 'storageState.json',
    /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
    actionTimeout: 0,
    /* Base URL to use in actions like `await page.goto('/')`. */
    baseURL: process.env.BASEURL,

    /* Collect trace when retrying the failed test. See v */
    trace: 'retain-on-failure',
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: process.env.BROWSER_NAME,
      use: {
        ...devices[process.env.DEVICE],
      },
    },

    // {
    //   name: 'firefox',
    //   use: {
    //     ...devices['Desktop Firefox'],
    //   },
    // },

    // {
    //   name: 'webkit',
    //   use: {
    //     ...devices['Desktop Safari'],
    //   },
    // },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: {
    //     ...devices['Pixel 5'],
    //   },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: {
    //     ...devices['iPhone 12'],
    //   },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: {
    //     channel: 'msedge',
    //   },
    // },
    // {
    //   name: 'Google Chrome',
    //   use: {
    //     channel: 'chrome',
    //   },
    // },
  ],

  /* Folder for test artifacts such as screenshots, videos, traces, etc. */
  // outputDir: 'test-results/',

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   port: 3000,
  // },
};

export default config;

**Comment by eatonjl** [comment]
Here's a link to see it in better formatting: https://wtools.io/paste-code/bCq1

**Comment by eatonjl** [comment]
@artemrudenko  Sorry for the late response. I am still trying to figure out how to get a repo set up. I did notice that it seems like this bracket is the one missing a closing bracket: **["grepInvert": null,**
I just want to make sure, is that opening bracket supposed to be there?

**Comment by eatonjl** [comment]
Does that also mean that it is from an area outside of the config that potentially has the issue?

**Comment by artemrudenko** [comment]
@eatonjl 
You can try to make your config as simple as possible and try to execute tests again. If all is ok, you can add config properties one by one in order to find one that cause this error.

**Comment by eatonjl** [comment]
> > Here's a link to see it in better formatting: https://wtools.io/paste-code/bCq1
> 
> @eatonjl I have tried to run sample tests with it and no any errors. If you aren't interested in seeing config in logs I can simply comment out logging for it. Or if you could please create a repo to reproduce issue.
> 
> Best Regards, Artem

I am trying to figure out how to simplify the config without breaking the tests. It is hard because this is code we are currently using for work and I don't wish to break too much. I tried to make a new test, but am having issues getting it to run in TeamCity (separate issue). From your message above, though, did you say you were able to run a test using my config file without any issues? I want to make sure I understood that correctly.

**Comment by artemrudenko** [comment]
> > > Here's a link to see it in better formatting: https://wtools.io/paste-code/bCq1
> > 
> > 
> > @eatonjl I have tried to run sample tests with it and no any errors. If you aren't interested in seeing config in logs I can simply comment out logging for it. Or if you could please create a repo to reproduce issue.
> > Best Regards, Artem
> 
> I am trying to figure out how to simplify the config without breaking the tests. It is hard because this is code we are currently using for work and I don't wish to break too much. I tried to make a new test, but am having issues getting it to run in TeamCity (separate issue). From your message above, though, did you say you were able to run a test using my config file without any issues? I want to make sure I understood that correctly.

I took your config, and run sample tests using it but without globalSetup: require.resolve('./global-setup').
After that I have added my custom global-setup.ts file and tried to run sample tests again.
In both cases there were no any errors.
I don't know, maybe your global-setup contains smth specific...

In any case as I said earlier I can simply remove config logging or make this configurable, just do let me know if this is ok for you or not.

Best Regards,
Artem

**Comment by eatonjl** [comment]
Again, sorry for the delayed reply (and thank you for all the help you are giving). Ultimately, if nothing works, I am fine removing config logging. What would entail making it configurable? I created a new test file and reset the Playwright config file back to default. The only thing I added was the entry for the reporter. I ran that in TeamCity and saw the same issue. So, I am still not seeing its connection to the config file. Here is what the default one looks like (including the addition of the reporter line):
import type { PlaywrightTestConfig } from '@playwright/test';
import { devices } from '@playwright/test';

/**
 * Read environment variables from file.
 * https://github.com/motdotla/dotenv
 */
// require('dotenv').config();

/**
 * See https://playwright.dev/docs/test-configuration.
 */
const config: PlaywrightTestConfig = {
  testDir: './e2e',
  /* Maximum time one test can run for. */
  timeout: 30 * 1000,
  expect: {
    /**
     * Maximum time expect() should wait for the condition to be met.
     * For example in `await expect(locator).toHaveText();`
     */
    timeout: 5000
  },
  /* Run tests in files in parallel */
  fullyParallel: true,
  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,
  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,
  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,
  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [
		['list'],
		['playwright-teamcity-reporter', { testMetadataArtifacts: 'test-results' }],
	],
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
    actionTimeout: 0,
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://localhost:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    trace: 'on-first-retry',
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'chromium',
      use: {
        ...devices['Desktop Chrome'],
      },
    },

    {
      name: 'firefox',
      use: {
        ...devices['Desktop Firefox'],
      },
    },

    {
      name: 'webkit',
      use: {
        ...devices['Desktop Safari'],
      },
    },

    /* Test against mobile viewports. */
    // {
    //   name: 'Mobile Chrome',
    //   use: {
    //     ...devices['Pixel 5'],
    //   },
    // },
    // {
    //   name: 'Mobile Safari',
    //   use: {
    //     ...devices['iPhone 12'],
    //   },
    // },

    /* Test against branded browsers. */
    // {
    //   name: 'Microsoft Edge',
    //   use: {
    //     channel: 'msedge',
    //   },
    // },
    // {
    //   name: 'Google Chrome',
    //   use: {
    //     channel: 'chrome',
    //   },
    // },
  ],

  /* Folder for test artifacts such as screenshots, videos, traces, etc. */
  // outputDir: 'test-results/',

  /* Run your local dev server before starting the tests */
  // webServer: {
  //   command: 'npm run start',
  //   port: 3000,
  // },
};

export default config;

**Comment by eatonjl** [comment]
Better format: https://wtools.io/paste-code/bCra

**Comment by eatonjl** [comment]
Very interesting - I only see that error when the tests are run in TeamCity. If I run it locally, that message does not show.

**Comment by artemrudenko** [comment]
> Very interesting - I only see that error when the tests are run in TeamCity. If I run it locally, that message does not show.

@eatonjl I think I have found the reason for that. 

From https://www.jetbrains.com/help/teamcity/service-messages.html#Service+Messages+Formats we can find the following:
> A single service message cannot contain a newline character inside it, it cannot span across multiple lines.

And since config is quite big and we stringify it with spaces=2 by default this causes our message to be multiline message.

Best Regards,
Artem

**Comment by eatonjl** [comment]
@artemrudenko I tried with the latest version and no longer se the error. I also no longer see the config logging. Was your latest change to simply remove it?

**Comment by artemrudenko** [comment]
> @artemrudenko I tried with the latest version and no longer se the error. I also no longer see the config logging. Was your latest change to simply remove it?
@eatonjl actually no, I have made several changes related to this, so now I am no longer prettifing it before output, also I have replaced double quotes with single quotes, and made config logging configurable, e.g. If you want to see it config in your logs you can add logConfig: true to your reporter config and it will output it.

Best Regards,
Artem

**Comment by eatonjl** [comment]
Great! That is wonderful news. Thank you for all the help, and for resolving the issue. You have been a great help and this tool is very useful. Have a great day!

**Comment by artemrudenko** [comment]
Thank you, @eatonjl, you are welcome, have a nice day too.

Please do let me know if you have any other issues or comments.

**Comment by eatonjl** [comment]
I should have checked this before ending our conversation, but I thought I should make you aware. I turned on the config logging, and I see this error when the tests run:
Error while parsing TeamCity service message: Property value not found (no "=" character). Valid service message has a form of "##teamcity[messageName name1='escaped_value' name2='escaped_value']" where escaped_value uses substitutions: '->|', [->|[, ]->|], |->||, newline->|n
For now, I will just keep the logging off as that seems to work. If you happen to find out how to get it to work with logging on, let me know and I would be happy to take a look. Thanks!

**Comment by artemrudenko** [comment]
> @artemrudenko I should have checked this before ending our conversation, but I thought I should make you aware. I turned on the config logging, and I see this error when the tests run: **Error while parsing TeamCity service message: Property value not found (no "=" character). Valid service message has a form of "##teamcity[messageName name1='escaped_value' name2='escaped_value']" where escaped_value uses substitutions: '->|', [->|[, ]->|], |->||, newline->|n** For now, I will just keep the logging off as that seems to work. If you happen to find out how to get it to work with logging on, let me know and I would be happy to take a look. Thanks!

Hi @eatonjl ,
Nice catch) Thank you. Please check with the latest version. Should be Ok now.

Best Regards,
Artem

**Comment by eatonjl** [comment]
@artemrudenko I am not sure if I did it wrong. I tried the install command and it said there was nothing to update. I tried the test again and got the same thing. Is there something more I need to do to update it this time?

**Comment by artemrudenko** [comment]
> @artemrudenko I am not sure if I did it wrong. I tried the install command and it said there was nothing to update. I tried the test again and got the same thing. Is there something more I need to do to update it this time?

Hi @eatonjl,
Could you please try to clean-up your node_modules folder before calling npm install?
Also after installation goto node_modules/playwright-teamcity-reporter/package.json and check version?

Best Regards,
Artem

**Comment by eatonjl** [comment]
OK, I got it updated, and it is working now. Thanks for fixing that! The new changes does make it mostly a big glob of text, so it is harder to use. But I assume that is what had to be done so it would play nice with TeamCity. I will see if I can find a way to copy it and paste it somewhere so I can break it up to more readable formats. Thanks! 
For an idea - it would be cool if the config log could be exported into a text file that you could post in TeamCity's artifacts page. That way it could have the newline breakups and still be obtainable. Just a thought. 
THANKS AGAIN!!

**Comment by artemrudenko** [comment]
> OK, I got it updated, and it is working now. Thanks for fixing that! The new changes does make it mostly a big glob of text, so it is harder to use. But I assume that is what had to be done so it would play nice with TeamCity. I will see if I can find a way to copy it and paste it somewhere so I can break it up to more readable formats. Thanks! For an idea - it would be cool if the config log could be exported into a text file that you could post in TeamCity's artifacts page. That way it could have the newline breakups and still be obtainable. Just a thought. THANKS AGAIN!!

You are welcome @eatonjl , in order to make it readable you can copy config string and than call JSON.stringify(copiedString, undefined, 2) for ex in browser console.

If you think that we need config as attachment I can look into this. Could you please crate new issue for this?

Best Regards,
Artem
